[클린코드 리액트 2기 김동호] 계산기 구현

psy082 opened the pull request:
안녕하세요 react 2기 김동호 입니다 :)
진행을 많이 하지 못했는데, 작게라도 시작해서 피드백을 받으려고 draft를 올렸습니다.

- 가장 최소한의 버전으로 사칙연산만 가능한 계산기를 구현했습니다.
- 이후 react와의 연결을 고려하여 이전까지의 입력을 축적한 상태에서 하나의 입력(digit or operator or ac)을 입력받았을 때 상태가 갱신되도록 구현하였습니다.

다음 과제 전까지 계속해서 버전을 발전시켜나가겠습니다. 

## 🎯 기능 요구사항

- [x] 2개의 숫자에 대해 덧셈이 가능하다.
- [x] 2개의 숫자에 대해 뺄셈이 가능하다.
- [x] 2개의 숫자에 대해 곱셈이 가능하다.
- [x] 2개의 숫자에 대해 나눗셈이 가능하다.
- [x] AC(All Clear)버튼을 누르면 0으로 초기화 한다.
- [x] 숫자는 한번에 최대 3자리 수까지 입력 가능하다.
- [x] 계산 결과를 표현할 때 소수점 이하는 버림한다.
- [x] 연산의 결과값이 `Infinity`일 경우 `오류`라는 문자열을 보여준다. (아이폰 참고)

psy082 committed:
init: 프로젝트 초기 설정

psy082 committed:
feat: Calculator 0.1 version 구현

- 단일 입력을 받아서 계산 결과를 출력함
- 괄호가 포함되지 않은 사칙연산이 가능한 버전

psy082 committed:
refactor: calculate 함수 내부 분기문 수정

Tanney-102 commented on .eslintrc.json:
```diff
@@ -0,0 +1,36 @@
+    "prettier"
+  ],
+  "rules": {
+    "prettier/prettier": "error",
```
`plugin:prettier/recommended`를 extends 하면 L21의 속성이 알아서 들어가는 걸로 알고 있습니다~
https://github.com/prettier/eslint-plugin-prettier#recommended-configuration

psy082 replied:
규칙에 대해서 충분히 찾지 않고 이곳 저곳에 있는 설정들을 조합해서 그렇습니다...
eslint 규칙들에 대해 기본적인 부분들은 숙지하고 있어야 겠네요..!

해당 규칙은 삭제했습니다!

Tanney-102 commented on src/Model/Calculator.test.js:
```diff
@@ -0,0 +1,37 @@
+    const result = Calculator(['999', '/', '3']).enter('=');
+
+    expect(result.value).toBe('333');
+  });
```
0으로 나누면 어떤 결과가 나오나요?

psy082 replied:
Infinity가 나옵니다!
구현 사항을 만족하기 위해 Infinity 값이 반환될 경우 '오류'라는 메세지가 출력되도록 기능을 추가했습니다!

Tanney-102 commented on src/Model/Calculator.js:
```diff
@@ -0,0 +1,73 @@
+  const postfix = [];
+  const operators = [];
+  let input;
+  while ((input = inputs.shift())) {
```
js array의 shift나 unshift는 O(n) 연산을 합니다. 따라서 의도하신 시간 복잡도를 달성하려면 따로 자료구조를 만드시는 편이 좋습니다.

Tanney-102 replied:
inputs에 꼭 변경이 있어야하는 게 아니라면 for of나 forEach 같은 걸 사용해 순회만 해도 충분할 것 같네요~

psy082 replied:
해당 부분은 전혀 생각하지 못했습니다 하하..
피드백 덕분에 자료 구조들의 시간복잡도 내용을 다시 살펴보게 되었습니다!

shift unshift를 사용하지 않고 for of 를 사용하도록 구현을  변경하였습니다 감사합니다 :)

Tanney-102 reviewed (changes requested):
안녕하세요 동호님!
미션 진행하느라 고생이 많으셔요~
멋진 알고리즘으로 계산기를 만들어주셨네요!
이 계산기가 리액트를 통해 UI로 표현되면 얼마나 더 좋게 느껴질지 기대가 됩니다!
아주 간단하게라도 좋으니 빠르게 시도해보시고 다시 리뷰요청 부탁드립니다!

psy082 committed:
refactor: calculate 함수 성능 개선

- shift, unshift -> push, pop으로 변경

psy082 committed:
refactor: 곱셈 연산 수정

- 지수 표기로 표현되는 큰 숫자의 경우에는 parseInt로는 제대로 변환되지 않으므로
  입력된 숫자 문자열을 숫자 타입으로 변환시키는 함수를 parseInt에서 Number로 변경

psy082 committed:
feat: 연산의 결과값이 Infinity일 경우 오류라는 문자열을 보여준다.

psy082 committed:
test: 곱셈 연산 테스트 케이스 추가

- 결과가 Infinity인 경우 오류가 발생하는 케이스 추가

psy082 committed:
test: 뺼셈 연산 테스트 케이스 추가

psy082 committed:
test: 덧셈 연산 테스트 케이스 추가

psy082 committed:
test: 나눗셈 연산 테스트 케이스 추가

- 기존에 분리되어 있던 소수점 이하 버림 케이스를 나눗셈 연산 케이스에 통합

psy082 committed:
feat: ac 입력 시 값을 0으로 초기화

psy082 committed:
feat: 숫자는 한번에 최대 3자리 수까지 입력 가능하다

- 숫자 입력 시 이전 입력이 숫자인 경우 새로운 숫자 입력이 되도록 구현
- 숫자 입력이 3자리 수가 넘어갈 경우 에러를 던지도록 구현

psy082 committed:
add: 연산자 이후에 연산자가 입력된 경우 연산자가 교체되도록 수정

psy082 committed:
chores: 불필요한 eslint rule 제거

- "prettier/prettier": "error" rule이 plugin:prettier/recommended에 포함되어 있으므로 제거

psy082 commented:
8가지 기능 요구 사항을 만족하는 계산기 ver 0.2를 만들었습니다.

데모 버전과 조금 다르게 만든 부분이 있습니다
- 데모 버전은 숫자 입력 전에 연산자를 입력하지 못하도록 했지만, 제가 구현한 계산기는 무조건 0이 있다는 전제하에 연산자를 허용합니다.
- 데모 버전은 연산자 다음 연산자를 입력하지 못하지만, 제가 구현한 계산기는 이전의 연산자를 새로운 연산자로 교체합니다.
 
추가적으로 개선이 필요한 점 
- 아직 음수 지원이 안되는 문제가 있어서 음수 지원 기능이 추가되어야 합니다.

음수 지원까지 기능을 추가한 후에 react 를 연결할 계획입니다.

피드백 주신 내용들에 감사합니다 :)

psy082 committed:
refactor: 덧셈 연산 시 숫자 문자열 변환 함수를 parseInt에서 Number로 변경

psy082 committed:
refactor: enter 함수 내부 분기문 수정

psy082 committed:
add: 다른 연산자 뒤에 입력되는 - 연산자는 음수로 허용하도록 수정

Tanney-102 reviewed (changes requested):
안녕하세요 동호님~
리뷰 반영해주신 부분 확인했습니다. ㅎㅎ
계산기를 지속적으로 개선하시는 모습이 인상적이네요!
다만 더 이상 다른 기능을 붙이기보다 우선 react를 사용한 UI 구현을 얼른 해보셨으면 좋겠어요.

1. 이 과정의 목적은 react를 사용하며 마주치는 다양한 상황을 경험하며 클린코드와 TDD에 대한 학습을 하는 것입니다.
2. 계산기 미션은 온보딩 과정으로 빠르게 리뷰 프로세스를 경험하는 것을 목적으로 합니다. 

따라서 너무 기능 고도화에 깊이 들어가는 것보다 목적에 맞는 학습을 진행하실 것을 권장드립니다.

psy082 committed:
add: negative이후의 연산자 입력이 무시되도록 수정

psy082 committed:
feat: react 화면 작성

- 기능 연결 없이 화면 구성만 작성

psy082 committed:
feat: Calculator와 react 화면 연결

psy082 commented:
안녕하세요 태은님!

기능 구현에 너무 깊이 빠졌던 것이 사실입니다...
바로 react와 연결을 시도했습니다!

사용자의 입력에 따른 계산 기능은 Calculator에 내장되어 있어서 
React 내부에서는 입력에 따라 Calculator를 업데이트 하도록만 했습니다.

여기서 질문은 제가 구현한 방식에서 각 버튼들이 별도의 컴포넌트로 나누어 질 필요가 있는 지 여부 입니다.

사실 다른 분들이 제출하신 과제를 같이 보았는데, 다른 분들은 대부분 적절하게 컴포넌트를 나누었다는 피드백을 받으신 것 같습니다.
그런데 다른 분들은 각각의 컴포넌트들이 내부에서 상태를 처리하거나, 아니면 외부에서 이벤트 핸들러를 받는 등의 역할이 분명한 것에 비해서 
지금 제가 작성한 리액트 엘리먼트들은 별도로 상태를 관리하지도 이벤트 핸들러를 받지도 않아서 
컴포넌트로 나누는 것이 과하지 않은가 하는 생각이 들었습니다.

이에 대해 태은님은 어떻게 생각하시는지 궁금합니다... 리액트 공부를 오랜만에 다시 시작하려고 하니 이런 기본적인 기준이 없다는 것이 다시금 드러나네요.. 

다음 과제부터는 react 사용에 좀 더 집중해보도록 하겠습니다.

추가로 질문이 있습니다!

react에서 상태를 관리하기 위한 모델을 만들 때, redux와 같은 도구를 사용할 경우에 모델링을 어떻게 해야 하는가? 하는 질문이 있습니다. 

제가 이번 과제를 할 때에 고민했던 부분은 다른 분들과 마찬가지로 계산을 처리하는 로직과 화면에 보여주는 부분을 분리하기 위해서는 어떻게 작성해야 하는가? 였습니다. 이전에  redux를 사용하는 경우에는 상태와 상태를 처리하는 로직을 무언가 도메인 동작을 응집력 있게 담게 되기 보다는 그 때 그 때의 상태 변화를 처리하도록만 작성하다가 전체적인 그림을 잃어버리게 되었던 것 같습니다. 그 때에서 아직 발전이 없어서 이번에는 아예 Calculator를 따로 구성하게 했는데요. 혹시 화면에서의 컴포넌트 작성 전 상태들을 관리하기 위해 도메인 모델링을 먼저 하는 경우에는 특히 react에서 사용하는 여러 상태 관리 도구들을 사용하는 경우에는 어떻게 작성하는 것이 좋을 지 궁금합니다!

psy082 committed:
add: exceedLimitOfDigitError 발생 시 에러 alert가 발생하도록 수정

psy082 committed:
test: 연산자 문자 변경으로 인한 테스트 케이스 수정

psy082 committed:
fix: - 입력 후 숫자 입력 시 음수가 되도록 수정

psy082 committed:
fix: -Infinity 또한 오류 문구를 출력하도록 수정

psy082 committed:
test: 연산자 뒤에 연산자가 입력될 때의 테스트 케이스 추가

Tanney-102 commented on src/App.jsx:
```diff
@@ -0,0 +1,49 @@
+};
+
+const App = () => {
+  const [calculator, setCalculator] = useState(Calculator());
```
```suggestion
  const [calculator, setCalculator] = useState(() => Calculator());
```

useState의 initialState가 식의 결과라면 위와 같이 함수로 전달하는 것이 권장됩니다~
https://reactjs.org/docs/hooks-reference.html#lazy-initial-state

Tanney-102 commented on src/App.jsx:
```diff
@@ -0,0 +1,49 @@
+
+const App = () => {
+  const [calculator, setCalculator] = useState(Calculator());
+  const handleClick = (e) => {
+    try {
+      setCalculator(calculator.enter(e.target.innerText));
+    } catch (e) {
+      if (e instanceof exceedLimitOfDigitError)
+        alert('숫자는 세 자리까지만 입력 가능합니다!');
+    }
```
'e'라는 변수가 한 함수 안에서 두가지 값을 의미하고 있네요.
가급적이면 변수 이름이 겹치지 않도록 해주시면 좋겠습니다!

Tanney-102 commented on src/Model/Calculator.js:
```diff
@@ -0,0 +1,117 @@
+
+export class exceedLimitOfDigitError extends Error {}
+
+export const Calculator = (inputs = ['0']) => {
```
Calculator 안에 있는 함수들의 크기가 너무 크다는 생각이 듭니다.
또 각 함수 안에서 일어나는 일들이 명령형 코드, 절자 지향적 코드로만 작성되고 있어요.
그래서 함수가 구체적으로 무슨 일을 하는지 빠르게 읽어내기 어렵습니다.
이렇게 크기가 큰 함수를 여러 작은 함수로 나눠보는 연습을 하셨으면 좋겠어요.
그리고 구체적인 함수 이름을 통해 어떤 일이 일어나고 있는지 표현해보셨으면 합니다. 
예를 들면 다음과 같은 식으로요!

```js
enter(value) {
  if (hasError(value)) { return Calculator('오류') }

  if (value === 'AC') { return empty() }
  if (value === '=') { return calculate(inputs) }

  if (value in OPERATOR) { return handleOperatorInput(inputs) }
  else { return handleNumberInput(inputs) }
}
```

Tanney-102 commented on src/Model/Calculator.js:
```diff
@@ -0,0 +1,117 @@
+      if (value in OPERATOR) {
+        // 연산자 이후에 연산자가 입력된 경우
+        if (recentInput in OPERATOR) {
+          if (recentInput === '-') {
+            const previousRecentInput = inputs[inputs.length - 2];
+            if (previousRecentInput in OPERATOR) return Calculator([...inputs]);
+          }
+
+          if (value === '-') {
```
recentInput === '-' 이 어떤 상황인지, value === '-' 은 어떤 상황인지 코드를 읽는 사람 입장에서는 금방 파악하기 어렵습니다. 가급적 의미있는 변수에 담아 사용해주세요.

ex)
```js
const previousRecentInput = inputs[inputs.length - 2];
const hasMinusAfterOtherOperator = recentInput === '-' && previousRecentInput in OPERATOR
if (hasMinusAfterOtherOperator) { return Calculator([...inputs]) }
```

Tanney-102 commented on src/Model/Calculator.js:
```diff
@@ -0,0 +1,117 @@
+      }
+
+      if (value in OPERATOR) {
+        // 연산자 이후에 연산자가 입력된 경우
```
코드를 설명하기 위해 주석을 사용하는 것은 그리 좋은 방법이 아닙니다.
특별한 예외 등이 있어 꼭 필요한 경우가 아니라면 코드 그 자체로 충분히 하고 있는 일을 설명하는 것이 좋습니다.
앞서 코멘트 드린 것처럼 작은 단위의 함수로 잘 나누고 의미있는 변수를 잘 활용하는 것만으로 주석의 필요성을 많이 줄일 수 있습니다.

Tanney-102 reviewed (approved):
동호님 안녕하세요!
UI쪽은 간결하게 잘 구현해주셔서 크게 피드백드릴만한 부분은 없었습니다!
소소한 코멘트 남겼으니 그것만확인 부탁드려요!

질문에 답변드립니다!

> 여기서 질문은 제가 구현한 방식에서 각 버튼들이 별도의 컴포넌트로 나누어 질 필요가 있는 지 여부 입니다.
...

사실 이 미션에서 구현하는 앱 자체가 너무 작은 규모라 어떻게 컴포넌트를 나누는게 좋은지에 대한 논의가 아주 큰 의미를 갖지는 않는 것 같아요. 말씀주신대로 지금 상태에서는 더 작은 단위로 나눌 필요가 크게 안느껴지긴합니다! 다른 분들이 받았던 피드백은 화면을 재사용할 수 있는 단위로 잘 나누고 각각을 컴포넌트로 구현해 보았다는 '행위'에 대한 부분이지 절대적인 기준이 있어서 적절했다는 의미는 아니었을 듯 싶습니다.(적어도 저는 이런 생각으로 리뷰했습니다~)

> react에서 상태를 관리하기 위한 모델을 만들 때, redux와 같은 도구를 사용할 경우에 모델링을 어떻게 해야 하는가? 하는 질문이 있습니다.
...

도메인에 경우 도구와 관련없이 작성되는 게 가장 좋지 않나 싶네요~
이미 동호님께서 도메인만을 담당하는 Calculator 함수를 잘 구현해주셨어요!
도메인은 툴에 관계없이 자기 할일만 잘하면 됩니다. 지금처럼 로컬에서 state로 사용하든지, redux 같은 상태관리 툴의 state로 사용하든지 도메인을 작성하는 일이 크게 달라지지는 않을 것 같습니다.

첫 온보딩 미션 고생많으셨습니다!
다음 미션을 위해 이 미션은 바로 머지하도록 하겠습니다!
